mrmustard/utils/parametrized.py

```diff
@@ -31,73 +32,28 @@ class Parametrized:
-                math.from_backend(value) and math.is_trainable(value)
-            )
-            constant = (
-                name + "_trainable" in kwargs
-                and kwargs[name + "_trainable"] is False
-                or (math.from_backend(value) and not math.is_trainable(value))
-            )
-            param_type = (
-                "symplectic"
-                if name.startswith("symplectic")
-                else "orthogonal"
-                if name.startswith("orthogonal")
-                else "euclidean"
-            )
-            if trainable:
-                if not math.from_backend(value) or (
-                    math.from_backend(value) and not math.is_trainable(value)
-                ):
-                    value = math.new_variable(value, kwargs[name + "_bounds"], owner + ":" + name)
-                self._trainable_parameters[param_type].append(value)
-            elif constant:
-                if not math.from_backend(value):
-                    value = math.new_constant(value, owner + ":" + name)
-                self._constant_parameters[param_type].append(value)
-            else:
-                name = "_" + name
-            self.__dict__[name] = value
-            self._param_names += [] if name.startswith("_") else [name]
+
+            # filter `{name}_trainable` or `{name}_bounds`` to become fields
+            # of the class as those kwargs are used to define the variables
+            if "_trainable" in name or "_bounds" in name:
+                continue
+
+            # convert into parameter class
+            is_trainable = kwargs.get(f"{name}_trainable", False) or math.is_trainable(value)
+            bounds = kwargs.get(f"{name}_bounds", None)
+            param_value = create_parameter(value, name, is_trainable, bounds, owner)
+
+            # dynamically assign variable as attribute of the class
+            self.__dict__[f"{name}"] = param_value
 
     @property
-    def trainable_parameters(self) -> Dict[str, List[Trainable]]:
-        r"""Returns the dictionary of trainable parameters, searching recursively in the object tree (for example, when in a Circuit)."""
-        if hasattr(self, "_ops"):
-            return {
-                "symplectic": math.unique_tensors(
-                    [p for op in self._ops for p in op.trainable_parameters["symplectic"]]
-                ),
-                "orthogonal": math.unique_tensors(
-                    [p for op in self._ops for p in op.trainable_parameters["orthogonal"]]
-                ),
-                "euclidean": math.unique_tensors(
-                    [p for op in self._ops for p in op.trainable_parameters["euclidean"]]
-                ),
-            }
-
-        return self._trainable_parameters  # default
+    def trainable_parameters(self) -> Sequence[Trainable]:
+        return [value for value in self.__dict__.values() if isinstance(value, Trainable)]
```

Comment:
in the current version this works also for circuits, which may be made out of other circuits and/or gates, so it needs to be able to recurse (in the current version this is done by the `for op in self._ops` part, though I had to hard code the name `_ops` into the Parameter class which is awkward because _ops is an attribute of the `Circuit` class)